### DISCUSSION TITLE
[Feature Request]: Build Service for firmwares with GitHub Actions

### DISCUSSION THREAD

**Discussion by hrueger** [josephdadams/TallyArbiter#339] [Ideas]
Just an idea: I believe that we already have a GitHub Actions workflow for some firmwares for embedded devices in place. Is that correct @MatteoGheza?

If so, can we implement some kind of build service where people just insert values for variables defined in the firmware file (like server ip, etc) and can then download their customized firmware binaries? I think this would make it a lot easier for people who are not confident to set up a dev environment.

Maybe we don't even need a server for that, we can probably do that by triggering GitHub Action workflows. But I'm not sure how we can get the artifacts to people.

**Comment by MatteoGheza** [comment]
IDK, maybe we can run a Workflow when someone comments on a discussion, but it requires a Github account.
PROs:
- External server not required
- Everything using the same infrastructure for building listeners, so finding bugs will be easier

CONs:
- Github account required for the user
- We can't build a GUI for configuring builds

I'd like to build something similar to https://github.com/benzino77/tasmocompiler/ since it has a beautiful GUI and we can run it using Docker.

Maybe we can build an "external, downloadable module" for TallyArbiter, that downloads an interface similar to tasmocompiler and tools required by it, without using Github Actions workflows.
PROs:
- Server infrastructure don't required
- We can build a custom UI for the "build service"

CONs:
- More possible bugs caused by the local user env

**Reply by hrueger** [reply]
Can't we make a [GitHub App](https://docs.github.com/en/developers/apps/getting-started-with-apps/about-apps) that the user authenticates to. Then, we have his GitHub API Token and can then fork the repo, create a new branch, update the variables and run the GitHub Action in his fork. 
All those API requests can happen on client side, thus we could just make this a static site hosted on GitHub Pages.

I think that this would be easier than creating a build system inside Electron...

What do you think?

**Reply by MatteoGheza** [reply]
Good idea. Doing this we can build a GUI inside GH Pages.
Is possible to run a Github App on the client-side, without a backend?

**Reply by hrueger** [reply]
Ah, I just checked the docs and unfortunately the private key is needed for the authentication...

**Reply by hrueger** [reply]
But we can just ask for a PAT (and provide docs on how to create it). This could be a one-time procedure as we can store it in localstorage. This is how [download-directory](https://github.com/download-directory/download-directory.github.io) does it.

**Reply by MatteoGheza** [reply]
Can we use the Github API without an app and authenticate the user on the client-side and use the API to fork?

**Reply by MatteoGheza** [reply]
> But we can just ask for a PAT (and provide docs on how to create it). This could be a one-time procedure as we can store it in localstorage. This is how download-directory does it.

Yes I mean this

**Reply by hrueger** [reply]
Yes, that's what I just replied. However, without an app we can't provide a nice "Authenticate with GitHub" UI flow and instead need to ask for a Personal Access Token which the user has to create and input manually

**Comment by josephdadams** [comment]
Can we adapt the various compiled clients to prompt the user for this information via a small internally hosted web server or something similar? So have the default behaviour be to boot up like any IOT device - you join your phone/computer to this temporary SSID network, browse to the web page and enter in some information like wifi, TA server IP/port, etc., and then it reboots and joins that. Then we do not need to have a workflow like this.

**Reply by hrueger** [reply]
You mean like with the WiFi configuration and the captive portal? That would actually make more sense than this idea...

**Reply by josephdadams** [reply]
Yes I was thinking the m5 stick client already does this but could be wrong. I know the Vmix m5stick client works this way so if needed we can look at that source code for an implementation.
